### PR TITLE
Since #23394 split relist and resync and most controllers don't need to

### PR DESCRIPTION
**Pull request by rrati**
periodically resync, most controllers have their resync intervals set to 0

### CODE REVIEW

**Comment by Random-Liu** [comment]
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/28250.

**Comment by timothysc** [comment]
@wojtek-t could you PTAL.

File: pkg/controller/daemon/controller.go [hunk 1]
```diff
@@ -207,7 +203,7 @@ func NewDaemonSetsController(podInformer framework.SharedIndexInformer, kubeClie
 }
 
 func NewDaemonSetsControllerFromClient(kubeClient clientset.Interface, resyncPeriod controller.ResyncPeriodFunc, lookupCacheSize int) *DaemonSetsController {
-	podInformer := informers.CreateSharedPodIndexInformer(kubeClient, resyncPeriod())
+	podInformer := informers.CreateSharedPodIndexInformer(kubeClient, controller.NoResyncPeriodFunc())
```
Comment (wojtek-t) [inline pkg/controller/daemon/controller.go]: @mikedanese - does daemon controller need periodic resyncs?

File: pkg/controller/petset/pet_set.go [hunk 2]
```diff
@@ -118,7 +116,7 @@ func NewPetSetController(podInformer framework.SharedIndexInformer, kubeClient *
 			},
 		},
 		&apps.PetSet{},
-		petSetResyncPeriod,
+		controller.NoResyncPeriodFunc(),
```
Comment (wojtek-t) [inline pkg/controller/petset/pet_set.go]: @bprashanth - does petset controller need periodic resyncs?

File: pkg/controller/podgc/gc_controller.go [hunk 3]
```diff
@@ -75,7 +75,7 @@ func New(kubeClient clientset.Interface, resyncPeriod controller.ResyncPeriodFun
 			},
 		},
 		&api.Pod{},
-		resyncPeriod(),
+		controller.NoResyncPeriodFunc(),
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller.go]: @caesarxuchao - does gc controller need periodic resyncs? (I think it doesn't).

Comment (caesarxuchao) [inline pkg/controller/podgc/gc_controller.go]: No, it doesn't

File: pkg/controller/petset/pet_set.go (hunk 2 above)
Comment (timothysc) [inline pkg/controller/petset/pet_set.go]: @smarterclayton ?

File: pkg/controller/daemon/controller.go (hunk 1 above)
Comment (timothysc) [inline pkg/controller/daemon/controller.go]: @mikedanese ping ?

**Comment by smarterclayton** [comment]
My understanding is that in practice almost everything needs to resync if they mutate a different resource than the one they are watching, because you can get arbitrarily long conflict chains on updates.

**Comment by smarterclayton** [comment]
The only controller that doesn't have to resync is one that watches the same resource as it is mutating and has no external mutations.

**Comment by wojtek-t** [comment]
@smarterclayton 

>
The only controller that doesn't have to resync is one that watches the same resource as it is mutating and has no external mutations.

I completely agree that controller needs to watch resources that it is mutating - I have no doubts about it. But I also think that this is basically the case for all our controllers.

However, I completely don't understand why you wrote that it should have "no external mutations". If you are modifying object X and someone else is also modifying it at the same time, and you gets a conflict, then you will observe the change done by this other component, and it will trigger you recomputing this object again.
Or did I misunderstood what you meant by "external mutations"?

**Comment by timothysc** [comment]
k, I'd like to redirect this to the specific questions that we're raised on PetSets and DaemonSets.  If the answer is "no" let's proceed.  Tests pass, performance gains are noticeable.

**Comment by smarterclayton** [comment]
Basically any resource where you can fail to "action" a particular resource that is not your primary watch, or experience ordering issues with another resources (A before B) that don't have infinite queueing needs to resync.  We don't want to turn resync off unless we are 100% sure of that.

The resync change actually broke two controllers in Origin like this yesterday, so I want to be over sure the change is correct for each of these.  Resync is cheaper than relist and erring on the side of caution is fine.  The bar for removing resync is a lot higher.

**Comment by smarterclayton** [comment]
@deads2k regarding resync

**Comment by timothysc** [comment]
@deads2k could you elaborate on any issue you had seen on the re-base, and more specifically how this change might affect you.  I'd like to steer clear of conjecture and deal with the concrete issue(s).

**Comment by deads2k** [comment]
> @deads2k could you elaborate on any issue you had seen on the re-base, and more specifically how this change might affect you. I'd like to steer clear of conjecture and deal with the concrete issue(s).

The change as currently proposed here won't affect the resync period on a shared informer which we've been using to consolidate the caches.

The changes to the reflector which separated out the relist (errors only) versus resync (periodic from cache), weren't wired through in a way that actually provided the resync functionality for the existing FIFO queue to back the reflector.  That caused all controllers that were wired against reflectors to be unable to correct for periodic failures that required retries.  (See https://github.com/openshift/origin/pull/9861 where we're putting humpty back together)

That's also used by the scheduler in kube, but I haven't investigated whether the scheduler can now end up ignoring pods needing scheduling.  It definitely changed the behavior and will prevent re-attempt behavior that existed in the past.

The explanation by @smarterclayton is good.  You watch thing X to do thing Y.  If Y fails, you  either retry forever in your own logic (this is now possible with all the workqueue changes we've made) or you try a few times, then fall back to a resync.  I definitely prefer to have a resync even if its on a long interval because it keeps us from having to be perfect.

**Comment by wojtek-t** [comment]
@smarterclayton @deads2k - thanks for explanations.

Regarding this PR - I think this is fine to switch of the resync for ReplicationController and ReplicaSet (all the decisions we made are based on RCs/RSs and pods and we are watching both). @smarterclayton - does it make sense?
Also GC controller doesn't need periodic resyncs.

I don't know about others, so it blocks me from approving this PR.

File: pkg/controller/daemon/controller.go (hunk 1 above)
Comment (mikedanese) [inline pkg/controller/daemon/controller.go]: Sorry for delay. So I'm torn here. A quick rescan doesn't show anything that is obviously broken without resyncs however I am worried about bugs. I would have the same concern with other controllers so if we have decided to do this with other controllers I guess we should do this with daemonsetcontroller as well. Also we are relatively early in the v1.4 merge window so this is a good time to do it.

Comment (mikedanese) [inline pkg/controller/daemon/controller.go]: And just to be pendantic, from api conventions:

> Functionality must be level-based, meaning the system must operate correctly given the desired state and the current/observed state, regardless of how many intermediate state updates may have been missed. Edge-triggered behavior must be just an optimization.

Why does only syncing on events not a violate of this convention? I'm not sure if this has been explicitly answered.

**Comment by rrati** [comment]
@wojtek-t @mikedanese It seems there are two paths here, and I would like to get this pr completed so we can all move on.  We can accept the PR as is and hash out any potential issues since it is early in the release cycle, or I can split into 2 PRs: 1 with ReplicationController, ReplicaSet, GC, 2nd with the rest.  That would allows us to merge the changes for the controllers we feel are safe and figure out how to handle the others.

Is there a preference?  If I don't hear back I'll split into 2 PRs.

**Comment by smarterclayton** [comment]
I am not comfortable with removing resync from any controllers, especially
without a strong justification in terms of scale numbers.  We took relist
out, that was billed as the biggest improvement, but resync is something
that needs a lot of justification.

**Comment by rrati** [comment]
What about if I add a toggle to disable setting the resync intervals to 0?  By default it would be enabled but if there are issues we can disable.

In past data we have seen about a 1 core improvement in the spikes when the re-syncs occur.  I'm working on getting current data for this change.

**Comment by wojtek-t** [comment]
It seems that @smarterclayton has very strong opinion on this one. So I'm reassigning to him.

**Pull request closed without merging**